proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -62,92 +62,149 @@ let userList = [User]() ❌
 
-The compiler offers existential types for all protocols except for ones that reference `Self` in a [non-covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) position or/and ones that have associated type requirements. However, that is incohesive as `Self` can be thought of as an associated type on its own, because it's can't be known solely from the context of a protocol; thus, giving `Self` and associated types different rules for synthesis of an existential type seems rather unintuitive. Furthermore, current semantics can promt library authors to avoid refining their protocols with other useful protocols in fear of their refined protocol not qualifying for the automatically generated existential type. Consider the following `Animal` protocol:
+The case of an associated type requirement that is known to have a predefined implementation reveals another downside to the status quo:
 
 ```swift
-protocol Identifiable {
+protocol P {
+  associatedtype A
+}
 
-  associatedtype ID : Hashable
+protocol Animal: Identifiable where ID == String {}
+extension Animal {
+  var name: String { id }
+}
+```
 
+Contrary to the intent, `Animal` is still assumed as "having an associated type requirement", which happens to be the only thing restraining the existential.
 
-  var id: ID { get }
+One more counterproductive aspect to the semantic inconsistency is how authors could be lead to avoid refining useful protocols in fear of their own protocol not qualifying for an existential type.
 
-}
+### Library Evolution
 
+Removing the type-level restriction would mean that adding defaulted requirements to a protocol is always both a binary- and source-compatible change, since it couldn't interfere with existing uses of the protocol.
 
-protocol Animal : Identifiable where ID == String {
+### Type-erasing Wrappers
 
-  var name: String { get }
-  
-  var speciesName: String { get }
-  
-  
-  var isPet: Bool { get }
+For convenience, libraries are often bound to vend custom type-erasing wrappers for commonly used protocols. Since the wrapper must evolve in sync with the protocol, this comes at the expense of resilience, code size and valuable time required to design, document, and maintain these ergonomic interfaces. While removing the restriction doesn't define away the need for manual containers like [`AnySequence`](https://developer.apple.com/documentation/swift/anysequence) and [`AnyHashable`](https://developer.apple.com/documentation/swift/anyhashable), it allows for them to be written in a way that's simpler, easier for the compiler to optimize, and ABI-compatible with future generalized existentials, by wrapping the unconstrained existential type instead of falling back on `Any` or boxing a value in a subclass or closure, and using private protocol extensions to wrap requirements that cannot be directly accessed on the existential:
 
-}
+```swift
+protocol Foo {
```

Comment:
I think we should avoid 'Foo' and 'Bar', because –– as Holly informed me during the writing of SE-293 –– it is actually an [offensive term](https://en.wikipedia.org/wiki/Foobar). Besides, I think we can find better examples with names and concepts that one is more likely to encounter in a real codebase. 

I think a guide for intuitive code examples is the [structured concurrency proposal's](https://github.com/DougGregor/swift-evolution/blob/eb149799aac93b3fb84a97adb2bb3db47179a21b/proposals/nnnn-structured-concurrency.md) dinner preparation examples.

Reply:
I agree we could (and I am going to) try harder on some examples, but I have mixed feelings on how reasonable that is for this particular one that shows a *general* approach to wrappers, or whether it is worth the time for some other rather subtle ones I would like to introduce in the "Detailed Design" section to cover the edge cases. These would most likely end up being an inseparable part of some large abstract interface if used in the wild.

Reply:
I think that an overarching set of protocols, compositions, or use cases in general, used consistently across the proposal could help in both consistency and comprehensibility. So in the case of Detailed Design, we could provide descriptions and then some notable examples to help tie the thoroughly-discussed behavior together. 

Of course, general examples are inherent in such an abstract feature –– which is not so much the case for the slightly more specific structured concurrency proposal. Structured concurrency is, of course, still an abstract concept; yet, the proposal authors did a great job of concentrating the concept into examples that showcase the its connection to a real-life task (cooking dinner). 

Likewise, property wrappers, through review comments, were identified to belong to some categories, such as utility, proxy, domain-bound (or something to that effect); based on these categories, we settled on some wrappers that were used throughout the text, which I think helps with coherence.

For this proposal, I think the theme is dynamism. Existential types can take the place of generics I think we should focus on the "correct" use cases. The reason why I initially included `AnyView` is because every user of SwiftUI will have probably experienced the live previews as well. These is a great example of how an existential type acts as a bridge between the strictly-typed world of SwiftUI and the dynamic environment of Xcode previews.